plugins/Login/Controller.php

```diff
@@ -220,6 +220,19 @@ protected function authenticateAndRedirect($login, $password, $urlToRedirect = f
         // remove password reset entry if it exists
         $this->passwordResetter->removePasswordResetInfo($login);
 
+        if (empty($urlToRedirect)) {
+            $referrer = Url::getReferrer();
+            $module = Common::getRequestVar('module', '', 'string');
+            // when module is login, we redirect to home...
+            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $referrer) {
+                $host = Url::getHostFromUrl($referrer);
+                // we only redirect to a trusted host
+                if ($host && Url::isValidHost($host)) {
```

Comment:
To increase security we could maybe do even `if ($host && Url::isValidHost($host) && $host === Url::getCurrentHost()) {` but this could be maybe tricky with proxies etc.

Reply:
isValidHost does `$host === Url::getCurrentHost()` basically, so should be ok IMO